### PR TITLE
`impl Rc::new_cyclic`

### PR DESCRIPTION
**Pull request by mental32**
References #75861

r? @Dylan-DPC

### CODE REVIEW

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -325,6 +325,40 @@ impl<T> Rc<T> {
         )
     }
 
+    /// Constructs a new `Rc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function retuns will result
```
Comment (andrewhickman) [inline library/alloc/src/rc.rs:329]:
```suggestion
    /// to upgrade the weak reference before this function returns will result
```

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -325,6 +325,40 @@ impl<T> Rc<T> {
         )
     }
 
+    /// Constructs a new `Rc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value. However, the weak reference may be cloned freely and
+    /// stored for use at a later time.
+    #[inline]
```
Comment (KodrAus) [inline library/alloc/src/rc.rs:332]: I think we could drop this `#[inline]` attribute off

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this @mental32!

This follows the same dance as [`Arc::new_cyclic`](https://github.com/rust-lang/rust/blob/master/library/alloc/src/sync.rs#L344-L393), which carefully avoids creating references to uninitialized data. Would you like to pull in some of the comments from `Arc::new_cyclic` into this `Rc::new_cyclic` function too so visitors can follow what's going on?

**Comment by mental32** [comment]
@KodrAus I've mad the suggested changes, feel free to review again :)

**Review by KodrAus — APPROVED** [review]
Thanks @mental32!

This looks good to me

**Pull request merged**